### PR TITLE
perf: use vectored writes

### PR DESCRIPTION
**Pull request by TestingPlant**
Currently submitting this draft to get an initial review. I won't be at my computer for a while to do any benchmarks or testing outside of checking if it compiles.
TODO: 
- [x] Benchmarks
- [ ] Check if it works on MacOS
- [ ] Avoid using channels due to performance costs
- [x] Consider adjusting write delay or alternatively writing after each tick - I've decided to not write after each tick because this could increase latency if it takes 51ms to write some packets at one tick, causing it to miss the next tick's write request.

### CODE REVIEW

File: crates/server/src/net.rs [hunk 1]
```diff
@@ -43,6 +46,8 @@ use valence_registry::RegistryCodec;
 use crate::{config, global};
 
 const DEFAULT_SPEED: u32 = 1024 * 1024;
+const MAX_VECTORED_WRITE_BUFS: usize = 8;
```
Comment (andrewgazelka) [inline crates/server/src/net.rs:49]: Currently, I think there is no reason this shouldn't be 2

Each game tick:
- one buf is sent from `Broadcast`
- one buf is sent from `Encoder`

Maybe we can exploit this somehow.

**Review by andrewgazelka — CHANGES_REQUESTED** [review]

**Comment by andrewgazelka** [comment]
image
this does not work on macOS for me 😔

File: crates/server/src/net.rs (hunk 1 above)
Comment (TestingPlant) [reply crates/server/src/net.rs:49]: If sending the data takes longer than a tick (such as when the tcp send queue is full) then there'll be more than 2 buffers in the channel.

**Comment by TestingPlant** [comment]
> this does not work on macOS for me 😔

That's weird. It works on Linux for me with io uring enabled and with it disabled.

**Comment by andrewgazelka** [comment]
> > this does not work on macOS for me 😔
> 
> That's weird. It works on Linux for me with io uring enabled and with it disabled.

hmmm I will try to debug. I will try running it in Docker and see if it is any different.

**Comment by TestingPlant** [comment]
I benchmarked this by timing the `send_data` function using `Instant::now()`, letting 250 bots join, waiting for all of them to finish joining, and then copying some of the output at the end. This probably isn't the best method, but I'm not aware of many better methods without creating an actual isolated benchmark outside of the code.

Original code from `main` branch with only timing added:
```
it took 6.334205ms to send_data 1929 bytes with no buffers and no vectored write
it took 6.334055ms to send_data 1929 bytes with no buffers and no vectored write
it took 6.332785ms to send_data 1929 bytes with no buffers and no vectored write
it took 6.331523ms to send_data 1929 bytes with no buffers and no vectored write
it took 6.330451ms to send_data 1929 bytes with no buffers and no vectored write
it took 6.330468ms to send_data 1929 bytes with no buffers and no vectored write
it took 6.330714ms to send_data 1929 bytes with no buffers and no vectored write
```

When `send_data` is replaced with the following code:
```rs
for bytes in bytes {
    let (result, _) = self.write.write_all(bytes.clone()).await;
    result?;
}
Ok(())
```
the results are:
```
it took 6.177458ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.179897ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.182388ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.184318ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.186832ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.188994ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.190817ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.193297ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.195602ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.197592ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.200378ms to send_data 9355 bytes with 6 buffers and no vectored write
it took 6.243289ms to send_data 9366 bytes with 6 buffers and no vectored write
it took 6.246547ms to send_data 9366 bytes with 6 buffers and no vectored write
```

When using this PR, the results are:
```
it took 2.093313ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.095398ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.097523ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.099838ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.102217ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.104327ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.106793ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.108915ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.111129ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.113584ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.1161ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.119443ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.121808ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.124015ms to send_data 9376 bytes with 6 buffers and vectored write
it took 2.126256ms to send_data 9376 bytes with 6 buffers and vectored write
```

I ran these multiple times and there were some outliers so I don't think my benchmark setup is perfect. In theory vectored writes should be faster though, and the benchmarks seem to prove this too.

**Comment by andrewgazelka** [comment]
> I benchmarked this by timing the `send_data` function using `Instant::now()`, letting 250 bots join, waiting for all of them to finish joining, and then copying some of the output at the end. This probably isn't the best method, but I'm not aware of many better methods without creating an actual isolated benchmark outside of the code.
> 
> Original code from `main` branch with only timing added:
> 
> ```
> it took 6.334205ms to send_data 1929 bytes with no buffers and no vectored write
> it took 6.334055ms to send_data 1929 bytes with no buffers and no vectored write
> it took 6.332785ms to send_data 1929 bytes with no buffers and no vectored write
> it took 6.331523ms to send_data 1929 bytes with no buffers and no vectored write
> it took 6.330451ms to send_data 1929 bytes with no buffers and no vectored write
> it took 6.330468ms to send_data 1929 bytes with no buffers and no vectored write
> it took 6.330714ms to send_data 1929 bytes with no buffers and no vectored write
> ```
> 
> When `send_data` is replaced with the following code:
> 
> ```rust
> for bytes in bytes {
>     let (result, _) = self.write.write_all(bytes.clone()).await;
>     result?;
> }
> Ok(())
> ```
> 
> the results are:
> 
> ```
> it took 6.177458ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.179897ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.182388ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.184318ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.186832ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.188994ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.190817ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.193297ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.195602ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.197592ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.200378ms to send_data 9355 bytes with 6 buffers and no vectored write
> it took 6.243289ms to send_data 9366 bytes with 6 buffers and no vectored write
> it took 6.246547ms to send_data 9366 bytes with 6 buffers and no vectored write
> ```
> 
> When using this PR, the results are:
> 
> ```
> it took 2.093313ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.095398ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.097523ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.099838ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.102217ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.104327ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.106793ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.108915ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.111129ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.113584ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.1161ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.119443ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.121808ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.124015ms to send_data 9376 bytes with 6 buffers and vectored write
> it took 2.126256ms to send_data 9376 bytes with 6 buffers and vectored write
> ```
> 
> I ran these multiple times and there were some outliers so I don't think my benchmark setup is perfect. In theory vectored writes should be faster though, and the benchmarks seem to prove this too.

Ok awesome :) 

let me look more into getting this to work on macOS and then we can merge it in :)

**Comment by andrewgazelka** [comment]
hmmmm this is so weird 9/10 times it seems to work

**Comment by andrewgazelka** [comment]
also get

image

**Comment by andrewgazelka** [comment]
image
at least writev is being used

**Comment by andrewgazelka** [comment]
also got 

```
  23:51:09.288 ERROR  Error sending packet: Invalid argument (os error 22) ... Os { code: 22, kind: InvalidInput, message: "Invalid argument" }
```

**Comment by andrewgazelka** [comment]
also

ERROR  Error sending packet: Bad address (os error 14) ... Os { code: 14, kind: Uncategorized, message: "Bad address"

I wonder if there is some weird memory mismatch issue or something

**Comment by andrewgazelka** [comment]
I kinda am starting to think there is a bug in `monoio` writev impl on macOS...

**Comment by andrewgazelka** [comment]
going to merge this in for now. i think main being up to date atm is better than it being stable

**Comment by andrewgazelka** [comment]
actually, now that I am looking at it I it, I think the bug affects both linux and macOS (as long as iouring is not supported)

**Comment by TestingPlant** [comment]
> also
> 
> ERROR Error sending packet: Bad address (os error 14) ... Os { code: 14, kind: Uncategorized, message: "Bad address"
> 
> I wonder if there is some weird memory mismatch issue or something

I also got this once a while ago but I couldn't reproduce the issue.

**Pull request merged**